Next r8 iteration/New submod framework

Booplicate opened the pull request:
Ideally should get updated after #9736 is merged.

Major refactoring of the submod framework due to Python 3 update.

### Changes:
- Loading procedure has been changed:
- - MAS now prevents arbitrary rpy loading unless debug mode is on
- - Submods now **must** be explicitly defined akin to spritepacks - via a header file, as well as being placed in the appropriate folder
- - - This will require a wiki update
- - - The header format is *very* close to the old-style submods, but is using json
- - - If any changes are made in the future, we will need to update:
- - - - `_SubmodSchema` (a dataclass-like structure for json validation)
- - - - `HEADER_VERSION` (version number of the json schema)
- - - - the `_Submod` class itself (if new fields were added, for example)
- - - Submod init, parsing, and loading happens at -999
- - - - All submod's AST nodes will be merged with the existing nodes, preserving order
- - - - All `early` blocks **will be executed at -999, but before submod init nodes**
- - - Submods must be places within `game/Submods` (sub-folders are supported)
- - - Header file must be in the root folder of the submod, it should be called `header.json`
- - Most of submod API is now private and should be used only by MAS code. **Submods should not access submods objects or interfere their loading in any way**
- - - `mas_submod_utils.isSubmodInstalled` and `mas_submod_utils.getSubmodDirectory` are public and safe to use by submods
- - - We can expose more of the API if submod-makers require more tools
- moved `mas_can_import` into `mas.can_import` py package
- - gutted the renpy version (`0imports.rpy`) and pointed everything to the py pkg
- moved `mas_threading` into `mas.threading` py package
- - pointed everything in the rpy (`zz_threading.py`) to the `mas.threading` py package

### Additions:
- new package - `pydantic`, a great tool for validation of arbitrary data
- `mas_submod_utils.require` - akin to NodeJS `require`, can import a python module from a filepath, handy for submods
- `_mas_loader` namespace (private, if some functionality is desired, it should be imported as a public function/class):
- - `include_module` - allows "including" a renpy module to load down the init pipeline with respect to the nodes init order
- - `do_modules_exist` - checks if a renpy module exists
- - `import_from_path` - import py modules from file system
- - - this is mostly for submods (see `mas_submod_utils.require`) to allow importing local python packages w/o interfering the global scope. For example when 2 submods need to import different versions of the same library
- `_mas_root` namespace:
- - `is_dm_enabled` - checks if debug mode is on
- added a `mas` root folder for mas-specific py packages

### Fixes:
- fixed a crash when a submod used a undefined screen for its setting pane

Booplicate committed:
wip: new submod framework

Booplicate committed:
smol fixes to types, comment out obsolete code

Booplicate committed:
control script loading

Booplicate committed:
req changes

Booplicate committed:
submod framework improvements

use `glob` to find submods
use f-strings
more validation
fixes

Booplicate committed:
sanity check setting pane exists

Booplicate committed:
store directory

Booplicate committed:
more wip submod framework

more typing
added `getSubmodDirectory`
improved err msgs
improved naming
wip submod loading

Booplicate committed:
allow loading submods

Booplicate committed:
implement `include_module`

The proper way of loading modules

Booplicate committed:
sort before loading

Booplicate committed:
support submods py-packs

Booplicate committed:
implement `import_from_path`

allows to dynamically import modules for submods

Booplicate committed:
import in submod utils

Booplicate committed:
update places where config was used

Booplicate committed:
more validation + fixes

Booplicate committed:
add `pydantic`

Booplicate committed:
use `pydantic` for json validation

Booplicate committed:
make class private + fixes

Booplicate committed:
smol

dreamscached commented:
Finally, an entry level increaser upgrade with great functionality! (Not sarcastic, this is awesome)

Booplicate committed:
dont promote using github

Booplicate committed:
cleanup

Booplicate committed:
handle ren_py

don't like the use of glob here, might need to switch to os.walk or smth else

Booplicate committed:
update module loading func

using the changes I added to renpy should give better performance during loading
especially when having a large number of submods installer

Booplicate committed:
fix typo that lead to nameerror

multimokia committed:
Merge remote-tracking branch 'upstream/content' into new_submod_framework

multimokia committed:
add .env support + fix indentation

multimokia commented on Monika After Story/game/zz_submods.rpy:
```diff
@@ -1,102 +1,472 @@
+            return value
+
+        @pydantic.validator("modules")
+        def validate_modules(cls, value, values):
+            if not value:
+                raise ValueError("Submod must define modules.")
+
+            # IMPORTANT: Sort in alpha
+            value = tuple(sorted(value))
+
+            submod_dir = values.get("directory", None)
+            if (
+                submod_dir is not None
+                and not _mas_loader.do_modules_exist(*(f"{submod_dir}/{m}" for m in value))
+            ):
+                raise ValueError(
+                    "One or more submod modules are missing: {}".format(
+                        ", ".join(map(lambda s: f"'{s}'", value))
+                    )
+                )
+
+            return value
```
Do we want to change how we pass these in for usability? I understand we wish not to allow `/*` patterns as that effectively restores prior behaviour of unspecified files, however I feel it can be cumbersome to need to retype the full paths i.e.:
```json
{
    ...,
    "modules": [
        "main",
        "subfolder/file1",
        "subfolder/file2",
        "subfolder/file3"
    ]
}
```

or did we want to swap to a more literal (in terms of how the author would see their filestruct) format i.e.
```json
{
    ...,
    "modules": [
        "main",
        "subfolder": {
            "file1",
            "file2",
            "file3"
        }
    ]
}
```

Booplicate replied:
Couldn't get this format to work, leaving the current variant for the time being.

multimokia commented on Monika After Story/game/screens.rpy:
```diff
@@ -3151,7 +3152,7 @@ screen submods():
                 xmaximum 1000
 
-                for submod in sorted(store.mas_submod_utils.submod_map.values(), key=lambda x: x.name):
+                for submod in submods:
```
We should add toggles here so players can enable/disable submods from loading

multimokia committed:
Add sample submod in testcases

multimokia commented:
Adding more things

All directory imports when being imported by `require` should implicitly be pointed to the `__init__.py` file if not provided as otherwise the import fails as "not found"

Furthermore, we should remove the SSL changes introduced by #8399 as it has been made redundant by renpy shipping SSL by default. We should aim to retain the self-updating-cert idea but that can still be done via certifi.

multimokia committed:
use renpy.config.gamedir here

multimokia committed:
never upload .disabled

multimokia commented:
Further:
- Our included certifi package is outdated and is also included by renpy.
- If we want to keep the auto cert updates, we should aim to update the one in the lib folder.
- Keeping our certifi package as-is will cause all submods making requests to not be able to have SSL connections.

multimokia committed:
fix renpy's "feature" of save """security"""

multimokia commented on Monika After Story/game/zz_submods.rpy:
```diff
@@ -1,102 +1,472 @@
+    HEADER_GLOB = "**/header.json"
+    SUBMODS_DIR = "Submods"
+
+    SANE_STR_PATTERN = re.compile(r'^[a-zA-Z_\u00a0-\ufffd][ 0-9a-zA-Z_\u00a0-\ufffd]*$')
```
Given the form of validation we use, this is an allow list.

Out of curiosity, why do we allow zero-width spaces and replacement characters? Given the fields that are validated by this pattern will be merged together to create the submod's update label, this will cause an invalid label to be generated, making it impossible for any submod using a name with these chars to be able to support version update scripts.

Unless there's a reason these are here, we should avoid supporting these chars.

dreamscached replied:
Maybe they could be stripped when making a label out of them?

Booplicate committed:
impl submod settings

Booplicate committed:
fix return type

Booplicate committed:
fix: use default with persistent

Booplicate committed:
add early_developer to loader

multimokia committed:
remove our ssl/certifi packages and add a todo

multimokia committed:
fix push/queue event

multimokia committed:
use constrained types for simple validation

NOTE: we do this because we want to easily export the _SubmodSchema
as a json schema. Using contypes applies these to the exported schema

multimokia committed:
submod schema to v1

Booplicate committed:
improve naming

Booplicate committed:
smol readability

Booplicate committed:
fix import for modules

Booplicate committed:
hide private field from schema

Booplicate committed:
impl platform check + refactor attr of Submod cls

Booplicate committed:
reduce nesting

Booplicate committed:
swap methods around

define on top what we use below, easier to read

Booplicate committed:
better naming

Booplicate committed:
disable broken submods + actually use settings

multimokia commented on Monika After Story/game/zz_submods.rpy:
```diff
@@ -71,6 +72,8 @@ init -1000 python in mas_submod_utils:
         coauthors: constrained_list(constrained_str(regex=LABEL_SAFE_NAME)) = []
         repository: str = ""
         priority: constrained_int(ge=-999, le=999) = 0
+        required_os: frozenset[str] = frozenset()
+        blacklist_os: frozenset[str] = frozenset()
```
Can we swap to [this approach](https://docs.pydantic.dev/latest/usage/types/#enums-and-choices) for required/blacklist OS value validation? This should remove the `validate_required_os` func, and simplify the blacklist one, along with enforcing appropriate values only via json schema

This will remove

Booplicate replied:
1. The list of possible values isn't static, at least I'm not aware of it. And I'd rather not hardcode values in our code.
2. We'd need at least one of the validators anyway to make sure they have no items in common

Any ideas?

multimokia replied:
Mmm, that brings me another concern actually, seeing as at a glance it's unclear what values are accepted here, I was anticipating a simple windows/mac/linux (to correspond with the `renpy.` checks respectively, does this account for platform versions?

If so do we really want/need to be that specific?

Re: point 2, I'd think we need the second one anyway as it presents some additional logic, so no harm in adding the logic for that there too.

Booplicate replied:
I actually was wondering if we want to allow checking versions/architecture, but it might be too much.
I know it will at least return `Linux`, `Darwin`, `Java`, `Windows`. There can be more. Also not sure if it supports android just yet, but at the same time don't want to limit it to only desktops.

>  I was anticipating a simple windows/mac/linux (to correspond with the renpy. checks respectively

Not a fan how renpy handles os checks. However if you think that's sufficient, I can change to that.

multimokia replied:
I think it's best if we base it on what we intend MAS to be supported on, so that means the primary 3 (also Java oh no. Embedded MAS?)
Likewise while we don't support android I see no reason to add OS config for android or iOS.

I don't much like the renpy one myself, however I don't think `platform` is much better either, seeing as it needs to be constantly updated. 
Based on how it acts with the current version in MAS r6, it doesn't even know win10 exists, let alone 11, this is bound to happen again as newer versions of Windows come out too. Therefore esp for versioning or anything else (which may even have a valid use case, but again it might be much for a submod specific thing, and might be better handled within submod logic itself) I don't think platform is a good solution either.

That said, I don't think it's _wrong_ to enforce win/lin/mac, seeing as those are the OSes we support. We have limited support for chromebook and only support them if they run Linux, likewise we don't support android, and I think it's safe to say any offshoot OS is not supported as well because we have zero ability to test for it. Likewsie it's likely people who use niche OSes like those understand this.

Booplicate replied:
> also Java oh no. Embedded MAS

Memory hog MAS 👀

> Likewise while we don't support android

I think that can happen eventually (like eventually eventually), so I want to design with that in mind,  the system should be expandable, ideally without adding extra code.

> Based on how it acts with the current version in MAS r6

It's old py2 code, apparently it's different now seeing its sources. So using `platform` is actually the most flexible way.

> That said, I don't think it's wrong to enforce win/lin/mac

In that case, is this list fine: `linux`, `mac`, `windows`? Also, is it `mac` or `darwin`?

multimokia replied:
Oh good, it uses the cmd `ver` command, and not the hardcoded value, okay that's much better so that is reliable.

If android support is a concern we'd need to check what platform returns on droid. As for iOS, I have no idea what it'd be

Booplicate replied:
> As for iOS, I have no idea what it'd be

A warning on the screen "get something better"

Booplicate committed:
change platform check

Co-Authored-By: multimokia <misty@multimokia.studio>

multimokia committed:
improving schemas + validation (WIP)

Booplicate committed:
fix for attr starting with underscore

Booplicate committed:
multiple fixes

Booplicate committed:
fix ci

Booplicate committed:
try ci w/o outdated rpycs

BoofOof32 commented:
Are there any blockers currently for this getting merged? It seems like it's mostly there aside from the 1 requested change (and even there, I can't seem to find which file needs changing, all of them appear resolved)

Booplicate commented:
#9493

ThePotatoGuy commented:
imports migration is on me, if theres others lmk.

ThePotatoGuy committed:
Merge branch 'content' into new_submod_framework

ThePotatoGuy commented:
should also wait for #10203

ThePotatoGuy committed:
Merge branch 'content' into new_submod_framework

ThePotatoGuy committed:
update gitignore for pycharm

ThePotatoGuy committed:
add renpy spec for renpy typing in py packages (slow wip)

ThePotatoGuy committed:
remove the children part

ThePotatoGuy committed:
move can import and threading into py packages, gut can import since its not as needed now

ThePotatoGuy committed:
drop the dataclass part since its not needed

ThePotatoGuy commented:
summary of latest changes:
* I pretty much killed most of the certifi/ssl `MASImport` object code. I've left the base `MASImport` in case that has uses.
* Added a new `mas` folder in python-packages. this should be the root folder for all of our python code, and we should move stuff to there as time allows. 
  * (If you're using pycharm, set the source root to `python-packages` and everything will work fine.)
  * moved `mas_threading` and `mas_can_import` into the `mas` py package (`mas.threading` and `mas.can_import`). see those files for examples on how to do a transition
  * added a `mas.renpyspec` package that should be a dummy class/set of classes so typing works for py packages that need access to renpy stuff (like the main renpy import or a specific `store`). I've only added barebones so add as necessary.

Booplicate commented:
> * added a `mas.renpyspec` package that should be a dummy class/set of classes so typing works for py packages that need access to renpy stuff

Did you try to import renpy directly? We should be able to use that. If we can't, then we should use typeshed vs dummy classes.

ThePotatoGuy committed:
delete renpyspec and import renpy

ThePotatoGuy commented:
removed renpyspec and just imported renpy. requires having the sdk as a library to have type hints.

ThePotatoGuy committed:
update ci env var

Booplicate committed:
fix: rm unused, this override moved to overrides.rpy

Booplicate committed:
fix: use old style annotation

Booplicate committed:
fix: renpy doesn't use Formatter anymore

see https://github.com/renpy/renpy/pull/5036 for details

Booplicate committed:
fix: use old style annotations

Booplicate committed:
build: bump renpy to 8.3.4

Booplicate committed:
fix: add debug print

Legendkiller21 committed:
possible docking station fix

Don't Scuff Open Inside

Legendkiller21 committed:
fix kissing transformation

Legendkiller21 committed:
submods tab description fix

substitutes your unsubstitute

Co-Authored-By: Totally a booplicate <53382877+Booplicate@users.noreply.github.com>

Booplicate committed:
rm pydantic

new version is not loadable by renpy, better to avoid it all together

Booplicate committed:
remove pydantic schemas, impl custom validation

Booplicate committed:
upd gitignore

Booplicate committed:
upd loader logic, better namings

Booplicate committed:
better naming

Booplicate committed:
move import_from_path to submods

Booplicate committed:
few renamings

Booplicate committed:
add debug screen

Booplicate committed:
rm debug print

Booplicate committed:
upd example header

Booplicate committed:
compatibility fix

Booplicate committed:
fix issue with calendar

I think this 'not' was added by mistake, in the birthday select event, after user picked a date, the calendar remained enabled mid dialogue, which caused flow interruption. New logic makes more sense: if the calendar was enabled, we enable it afterwards, otherwise it remains disabled

Booplicate committed:
fix crash during nou due to unexpected int to float conversion

Booplicate committed:
fix more int/float div errors in nou

Booplicate committed:
fix freeze in chess, add a TODO

Booplicate committed:
rm unnecessary lock

deque is already thread safe

Booplicate committed:
rm certifi upd logic, rm masimport for certifi and ssl

this is handled by renpy 8, we dont need to worry about it now

Booplicate committed:
fix loading quicksave in chess

Booplicate committed:
upd gitignore

Booplicate committed:
remove unicode references, it's all str now

Booplicate committed:
remove long references, it's all int now

Booplicate committed:
impr code for managing savelocations

added comments from the OG function
added a runtime check to make sure there's always only one savelocation
cleanup

Booplicate committed:
apparently this bug is gone, remove unnecessary show

Booplicate committed:
finish some py3 TODO/FIXME for islands

Booplicate committed:
finish some more py3 TODO/FIXME in utils, chess, and screens

Booplicate committed:
rm old deprecated transforms

Booplicate committed:
rm priority from submods

this will be replaced with proper dependency resolution

Booplicate committed:
smol rename, rm global path patch, add TODOs

Booplicate committed:
impl git updater

Booplicate committed:
add schema for submod updater

Booplicate committed:
fix: jarg unittest

Booplicate committed:
refactor compareVersionLists

1. this version works in one pass
2. can use any sequence, not just lists
3. no list.append - faster
4. shorter and simpler, avoids many previous edge cases

Booplicate committed:
test: add unittests for new mas_utils.compare_versions

Booplicate committed:
docs: better docstring

Booplicate committed:
next iteration of the git updater

Booplicate committed:
first integration of submod updater

Booplicate committed:
refactor update scripts into update hooks

Booplicate committed:
rm version_updates, not used anymore

Booplicate committed:
only allow one update hook per version

Booplicate committed:
add installation hooks + fix bugs with update hooks

Booplicate committed:
add installation hooks (don't enable them yet)

Booplicate committed:
improve submods screen, better handle dependency check

Booplicate committed:
smol rename cleanup

Booplicate committed:
better UI for submod menu

Booplicate committed:
add logic for running installation hooks

Booplicate committed:
smol fix: use cm for reading moni file

Booplicate committed:
improve submod screen

Booplicate committed:
add topological sort for submods

Booplicate committed:
disable submods by default

Co-Authored-By: multimokia <misty@multimokia.studio>

Booplicate committed:
improve handling of disabled submods; try to resolve circular dependency

Booplicate committed:
bump timeout for git to 30s

this way even big submods should be able to checkout

ThePotatoGuy committed:
a setup script for copying over files and preparing a renpy folder

ThePotatoGuy committed:
Merge branch 'new_submod_framework' of github.com:Monika-After-Story/MonikaModDev into new_submod_framework

ThePotatoGuy committed:
initializing dev and scripts.rpa required for now

ThePotatoGuy committed:
mod package is back and new updater urls

ThePotatoGuy committed:
remove game lines in build

ThePotatoGuy committed:
add mac

ThePotatoGuy committed:
take out mac from the builds

ThePotatoGuy committed:
build, s3 scripts updated with mac and mod distributions

ThePotatoGuy committed:
actions test + keep scripts but use the poemgame hack

ThePotatoGuy committed:
try no opengl setup

ThePotatoGuy committed:
embarrasing cmd arg

ThePotatoGuy committed:
correct pckage, also try with compile option

ThePotatoGuy committed:
ok fine no updatre flag

ThePotatoGuy committed:
remove compile again

ThePotatoGuy committed:
interpolate exprs on so we get py eval for quips

ThePotatoGuy committed:
delay thanks quip?

ThePotatoGuy committed:
add CI only build key to change update behavior

multimokia commented on utils/mas_dev_setup:
```diff
@@ -0,0 +1,70 @@
+#   <base ddlc folder> - the folder containing DDLC (expects game/ and DDLC.exe)
+#   <dev (repo) folder> - the folder containing git/repo code (dev env)
+#       - expects either "Monika After Story/game/*.rpy" or "game/*.rpy"
+#   <rpy folder> - can be anything - set it to whatever renpy will see
```
`rpy folder` is misleading, why not call it `out` or `dist folder`? `rpy folder` could imply SDK

Also, should we not have this factor in the SDK to build a binary for the specific renpy version? If this is building just for 6 I don't think this is useful especially for users looking to get into development

multimokia replied:
Right, but if you're pointing to a DDLC install, this script expects the install is running with renpy 8 already

It doesn't actually build the rpys into rpycs, leaving that part to the DDLC install
but if the DDLC install is pure from the official source, it's running renpy 6

ThePotatoGuy replied:
yeah this isn't for that, its for setting up the folder so you can launch it with the r8 sdk (so the r8 sdk builds it). also the script doesnt copy renpy/libs/exe from the ddlc folder so it cant actually run by itself.

ThePotatoGuy replied:
mm u did this in the end so i think this is resolved

ThePotatoGuy committed:
rename rpy_folder to out_folder

multimokia committed:
fix(dev-setup-script): Don't copy scripts.rpa as renpy8 will not run with it

multimokia committed:
Merge branch 'content' of github.com:Monika-After-Story/MonikaModDev into new_submod_framework

# Conflicts:
#	Monika After Story/game/overrides.rpy

multimokia committed:
fix(chess): Resolve an issue where the `subprocess.STARTUPINFO` annotation is not recognized

Booplicate committed:
feat: further work on new submods

  fetch and update from the submdos screen
  cleanup code
  update notifs
  improve regex for validating names
  improved updater

Booplicate committed:
feat: add tooltips to submod screen; add install button

Booplicate committed:
feat(submods): custom notif screen;only check enabled submods for updates

Booplicate committed:
refactor(submods): improve naming

Booplicate committed:
refactor(submods): use match case

Booplicate committed:
perf(submods): cache submods list for submods screen

Booplicate committed:
fix(submods): guard an attribute with a lock

Booplicate committed:
refactor(submods): modify init order

Booplicate committed:
refactor(functionplugin): cleanup and optimise

Booplicate committed:
feat(functionplugin): always handle errors, but always allow call/jump from py code

ThePotatoGuy commented:
OK where were we at with this again? i spent like an hour trying to find where i made the v13 links and i see it was here
